From 1809c974bdab9ca3d7ad2e1a4480e53b53bc65f6 Mon Sep 17 00:00:00 2001
From: Thales Elero Cervi <thaleselero.cervi@windriver.com>
Date: Wed, 21 Sep 2022 16:48:54 -0300
Subject: [PATCH] Fixing helm release hooks and weights

The relation of dependency for cinder, nova and keystone releases resources is
not working with helmv3 since several jobs have post-install hooks and are
dependencies of other jobs and deployments that have no hooks.

The jobs/deployments without hooks are deployed during an installation
phase that is never complete since the dependency jobs are hooked to be
deployed on post-install phase.

This change includes helm-hooks to a few templates in nova, cinder and keystone.
The weights will define the order each one will be deployed.

Signed-off-by: Thales Elero Cervi <thaleselero.cervi@windriver.com>
Change-Id: I74dd271d065a7b4668845accae7476d5cbd7d363
[ upversioned of changes to openstack-helm for epoxy release ]
Signed-off-by: cfigueir <carlos.figueiredofelipe@windriver.com>
[ Merged similar patches under new OSH package]
Signed-off-by: Daniel Caires <DanielMarques.Caires@windriver.com>
---
 cinder/templates/deployment-api.yaml           | 4 ++++
 cinder/templates/deployment-scheduler.yaml     | 4 ++++
 cinder/templates/deployment-volume.yaml        | 4 ++++
 keystone/templates/secret-credential-keys.yaml | 3 ++-
 keystone/templates/secret-fernet-keys.yaml     | 3 ++-
 nova/templates/job-bootstrap.yaml              | 5 +++++
 nova/templates/job-cell-setup.yaml             | 4 ++++
 7 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/cinder/templates/deployment-api.yaml b/cinder/templates/deployment-api.yaml
index 571d71029..8d1a753d4 100644
--- a/cinder/templates/deployment-api.yaml
+++ b/cinder/templates/deployment-api.yaml
@@ -27,6 +27,10 @@ metadata:
   name: cinder-api
   annotations:
     {{ tuple $envAll | include "helm-toolkit.snippets.release_uuid" }}
+{{- if .Values.helm3_hook }}
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-weight: "1"
+{{- end }}
   labels:
 {{ tuple $envAll "cinder" "api" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 4 }}
 spec:
diff --git a/cinder/templates/deployment-scheduler.yaml b/cinder/templates/deployment-scheduler.yaml
index f43a68c23..d01200962 100644
--- a/cinder/templates/deployment-scheduler.yaml
+++ b/cinder/templates/deployment-scheduler.yaml
@@ -27,6 +27,10 @@ metadata:
   name: cinder-scheduler
   annotations:
     {{ tuple $envAll | include "helm-toolkit.snippets.release_uuid" }}
+{{- if .Values.helm3_hook }}
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-weight: "2"
+{{- end }}
   labels:
 {{ tuple $envAll "cinder" "scheduler" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 4 }}
 spec:
diff --git a/cinder/templates/deployment-volume.yaml b/cinder/templates/deployment-volume.yaml
index 240204256..eb9b17bfd 100644
--- a/cinder/templates/deployment-volume.yaml
+++ b/cinder/templates/deployment-volume.yaml
@@ -29,6 +29,10 @@ metadata:
   name: cinder-volume
   annotations:
     {{ tuple $envAll | include "helm-toolkit.snippets.release_uuid" }}
+{{- if .Values.helm3_hook }}
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-weight: "2"
+{{- end }}
   labels:
 {{ tuple $envAll "cinder" "volume" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 4 }}
 spec:
diff --git a/keystone/templates/secret-credential-keys.yaml b/keystone/templates/secret-credential-keys.yaml
index 8a2c5eb5b..307bb72bd 100644
--- a/keystone/templates/secret-credential-keys.yaml
+++ b/keystone/templates/secret-credential-keys.yaml
@@ -21,7 +21,8 @@ metadata:
   name: keystone-credential-keys
 {{- if .Values.helm3_hook }}
   annotations:
-    "helm.sh/hook": pre-install
+    "helm.sh/hook": pre-install,post-upgrade
+    "helm.sh/hook-weight": "-6"
 {{- end }}
 type: Opaque
 data:
diff --git a/keystone/templates/secret-fernet-keys.yaml b/keystone/templates/secret-fernet-keys.yaml
index 8af097309..a7eddd14e 100644
--- a/keystone/templates/secret-fernet-keys.yaml
+++ b/keystone/templates/secret-fernet-keys.yaml
@@ -22,7 +22,8 @@ metadata:
   name: keystone-fernet-keys
 {{- if .Values.helm3_hook }}
   annotations:
-    "helm.sh/hook": pre-install
+    "helm.sh/hook": pre-install,post-upgrade
+    "helm.sh/hook-weight": "-6"
 {{- end }}
 type: Opaque
 data:
diff --git a/nova/templates/job-bootstrap.yaml b/nova/templates/job-bootstrap.yaml
index 2aa62d21d..cddc74439 100644
--- a/nova/templates/job-bootstrap.yaml
+++ b/nova/templates/job-bootstrap.yaml
@@ -30,6 +30,11 @@ metadata:
   name: {{ $serviceAccountName | quote }}
   labels:
 {{ tuple $envAll "nova" "bootstrap" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 4 }}
+  annotations:
+{{- if .Values.helm3_hook }}
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-weight: "2"
+{{- end }}
 spec:
   backoffLimit: {{ $backoffLimit }}
   template:
diff --git a/nova/templates/job-cell-setup.yaml b/nova/templates/job-cell-setup.yaml
index f9d2e084a..0fd99e023 100644
--- a/nova/templates/job-cell-setup.yaml
+++ b/nova/templates/job-cell-setup.yaml
@@ -25,6 +25,10 @@ metadata:
   labels:
 {{ tuple $envAll "nova" "cell-setup" | include "helm-toolkit.snippets.kubernetes_metadata_labels" | indent 4 }}
   annotations:
+{{- if .Values.helm3_hook }}
+    helm.sh/hook: post-install,post-upgrade
+    helm.sh/hook-weight: "1"
+{{- end }}
     {{ tuple $envAll | include "helm-toolkit.snippets.release_uuid" }}
 {{ tuple "nova_cell_setup" $envAll | include "helm-toolkit.snippets.custom_job_annotations" | indent 4 }}
 spec:
-- 
2.34.1
